[MKC-761] Bootloader reloading (non-AVR boards) tutorial (#738)
arduino/docs-content: content/tutorials/generic/bootloader-reloading/bootloader-reloading.md

---

> title: Burn the bootloader on non-AVR boards
> description: Learn how to burn the bootloader on non-AVR boards.
> difficulty: Advanced
> featuredImage: 'clip'
> ---
> 
> In this tutorial, we will learn how to burn the bootloader on non-AVR boards using another Arduino board. The bootloader is a small piece of software that allows uploading sketches onto the Arduino board. It comes preprogrammed on the microcontrollers on Arduino boards. Whether the bootloader has been corrupted or intentionally has been removed, it can be restored by burning (also called, flashing or programming) a new bootloader to the board.

Reviewer: ```suggestion
title: Burn the bootloader on SAMD boards
description: Learn how to burn the bootloader on SAMD boards.
difficulty: Advanced
featuredImage: 'clip'
---

In this tutorial, we will learn how to burn the bootloader on boards that use the ATSAMD21G18 microcontroller using another Arduino board. The bootloader is a small piece of software that allows uploading sketches onto the Arduino board. It comes preprogrammed on the microcontrollers on Arduino boards. Whether the bootloader has been corrupted or intentionally has been removed, it can be restored by burning (also called, flashing or programming) a new bootloader to the board.
```

The instructions in this tutorial are specific to boards with the ATSAMD21 microcontroller. Since there are many non-AVR microcontrollers, the description of it being a tutorial for non-AVR boards was misleading

>**Note:** If you want to burn the bootloader on AVR boards such as an UNO, Mega, and classic Nano you can check out [these instructions](https://support.arduino.cc/hc/en-us/articles/4841602539164-Burn-the-bootloader-on-UNO-Mega-and-classic-Nano-using-another-Arduino).

## Requirements

> - An extra Arduino board that runs at 3.3V to use as the programmer.
> 
> >**Note:** Certain Arduino boards can't be used with the sketch that converts it to a programmer. It is possible to use an Arduino board that runs at 5V as the programmer, but you will need to use level shifting circuitry on the programming lines to avoid exposing the target board to 5V logic levels, which would damage it.
> 
> | Working | Untested | Not working |
> | ----------- | ----------- |----------- |
> | SAMD architecture boards | AVR architecture boards | Nano 33 BLE |
> 

Reviewer: ```suggestion
- An extra SAMD-based Arduino board to use as the programmer.
```

In the context of a forum thread, it might have been reasonable to start a discussion about which boards might be used as a programmer, but in the context of a tutorial, that is inappropriate. The tutorial should be focused on a known working procedure the reader can follow.

I do think it would be worthwhile, and also quite a fun assignment, to take the time to test whether other official Arduino boards might be used as a programmer. I verified that the example sketch is still failing to compile for the **Nano 33 BLE** board, so that is out, but it does compile for **Nano RP2040 Connect** and **Portenta H7 (M7 core)**, which I think are both interesting programmer candidates. It also compiles for **Mega** (other AVR boards don't have enough memory for the sketch), but I think that is a less valuable programmer candidate for the purposes of the tutorial due to the need to do logic level translation between the 5 V Mega and the 3.3 V SAMD board.

If other boards are verified to work, they could be added to a list here (I would still avoid adding a list of non-working or untested boards).

- Click the **"Install"** button.
- Wait for the installation to finish.
- Click the **Close** button.
- Select **File > Examples > Adafruit DAP library > samd21 > flash_from_SD** from the Arduino IDE's menus.
- Change this line: ```#define SD_CS 4``` according to the Arduino pin connected to the SD CS pin. 

    If your board has a built-in SD slot (e.g., [MKR Zero 9](https://store.arduino.cc/products/arduino-mkr-zero-i2s-bus-sd-for-sound-music-digital-audio-data)), then you can change this line:

    ```
    (!SD.begin(SD_CS)) { 
    ```

    to:

    ```
    if (!SD.begin()) {
    ```

- Select the correct board from the **Tools > Board** from the Arduino IDE's menus.
- Select the correct port from the **Tools > Port** from the Arduino IDE's menus.
- Select **Sketch > Upload** from the Arduino IDE's menus.
- Wait for the upload to finish successfully.
- Unplug the programmer Arduino board from your computer.
- Plug the SD card into the SD slot connected to your Arduino board.
- Connect the programmer Arduino board to the target Arduino board as follows:

    | Programmer  | Target      |
    | ----------- | ----------- |
    | VCC         | +3V3        |

>     | 10          | SWDIO       |
>     | 9           | SWCLK       |
>     | GND         | GND         |
>     | 11          | RESETN      |

Reviewer: ```suggestion
    | 12          | SWDIO       |
    | 11          | SWCLK       |
    | GND         | GND         |
    | 9           | RESETN      |
```

They changed the pins in the example sketch since the time I wrote that forum post:

https://github.com/adafruit/Adafruit_DAP/blob/1.7.4/examples/samd21/flash_from_SD/flash_from_SD.ino#L6-L8

    SWD pads on MKR boards other than MKR 1000:

>     ![](assets/SWDpadsMKR1000.png)

Reviewer: The filename is inappropriate. This image shows the pads on the MKR boards other than the MKR 1000, so a more accurate filename would be something like `SWDpadsMKR.png`

    MKR1000 SWD header pinout:

    ![](assets/SWDMKR1000header.png)

    Nano 33 IoT SWD pads:

    ![](assets/SWDpadsNano33IoT.png)

- Plug the USB cable of the programmer Arduino board into your computer.
- Select **Tools > Serial Monitor** from the Arduino IDE's menus. You should now see the Serial Monitor output showing the target board detected, and the bootloader file flashed to it successfully.
- Unplug the programmer Arduino board from your computer.
- Disconnect the programmer Arduino board from the target Arduino board.

> ## Alternatives
> These are some alternatives to the "Adafruit DAP" method described above.
> 
> ### Using a CMSIS-DAP debug probe as the programmer
> 
> If you have a CMSIS-DAP compliant debug probe, you can just do this instead:
> 
> - Connect the debug probe to your Arduino board.
> - Select **Tools > Programmer > Atmel EDBG** from the Arduino IDE's menus.
> - Select **Tools > Burn Bootloader** from the Arduino IDE's menus. - The "Burn Bootloader" process should now finish successfully.
> 
> You can use [this little open-source debugger](https://www.tindie.com/products/ataradov/cmsis-dap-compliant-swd-debugger/).
> 
> ### Using a J-Link as the programmer
> 
> Segger J-Link debug probes (e.g., J-Link EDU Mini) can be used with the Adalink software:
> https://github.com/adafruit/Adafruit_Adalink.
> 
> This is a fairly complex procedure, so we recommend against this option unless you already own a J-Link and don't have the supplies on hand for one of the other options.

Reviewer: ```suggestion
```

This is another case where I think it was maybe reasonable to mention in the context of a forum thread, but not appropriate for a tutorial. It would be worth writing separate tutorials for each of these alternative methods (though I found the software setup for the "Adalink" procedure to be very challenging when I performed it some years ago), then linking to each of the specific tutorials from an entry page so the reader can select the best method according to what hardware they happen to have on hand (e.g., if they have a CMSIS-DAP debug probe, that will be the easiest option by far).

---
title: Burn the bootloader on SAMD boards
description: Learn how to burn the bootloader on SAMD boards.
difficulty: Advanced
featuredImage: 'clip'
---

In this tutorial, we will learn how to burn the bootloader on boards that use the ATSAMD21G18 microcontroller using another Arduino board. The bootloader is a small piece of software that allows uploading sketches onto the Arduino board. It comes preprogrammed on the microcontrollers on Arduino boards. Whether the bootloader has been corrupted or intentionally has been removed, it can be restored by burning (also called, flashing or programming) a new bootloader to the board.

Below you will find two different approaches to burning the bootloader:
- The first approach uses an SD card and should be used when wanting to flash a different board than the ones bundled with the "flash_MKR_bootloaders" (check out the [Github repository](https://github.com/adafruit/Adafruit_DAP) for more information).

> - The second approach does not require a SD card and should be used by users who want the easiest solution.

Reviewer: ```suggestion
- [The second approach](#no-sd-card) does not require an SD card and should be used by users who want the easiest solution.
```

- Link to the relevant section of the page to make it easier for the reader to find it.
- Grammar nitpick.
  "SD" is pronounced "ess dee", and the choice of "a" vs "an" is based on pronunciation rather than spelling.

>**Note:** If you want to burn the bootloader on AVR boards such as an UNO, Mega, and classic Nano you can check out [these instructions](https://support.arduino.cc/hc/en-us/articles/4841602539164-Burn-the-bootloader-on-UNO-Mega-and-classic-Nano-using-another-Arduino).

## SD card
### Requirements
- An extra SAMD-based Arduino board to use as the programmer. 

> - An SD slot. This could be built into your Arduino board (e.g., [MKR Zero](https://store.arduino.cc/products/arduino-mkr-zero-i2s-bus-sd-for-sound-music-digital-audio-data?_gl=1*17dcyg9*_ga*MjEyMzQ2MjgwOC4xNjY1NjUyNTY3*_ga_NEXN8H46L5*MTY3MTYyNzMzMS4xNjEuMS4xNjcxNjI5ODA2LjAuMC4w)), a shield (e.g., [MKRSD Proto](https://store.arduino.cc/products/mkr-sd-proto-shield?_gl=1*1xs1eol*_ga*MjEyMzQ2MjgwOC4xNjY1NjUyNTY3*_ga_NEXN8H46L5*MTY3MTYyNzMzMS4xNjEuMS4xNjcxNjMwMzgxLjAuMC4w)), or one of the common SD modules.

Reviewer: ```suggestion
- An SD slot. This could be built into your Arduino board (e.g., [MKR Zero](https://store.arduino.cc/products/arduino-mkr-zero-i2s-bus-sd-for-sound-music-digital-audio-data)), a shield (e.g., [MKR SD Proto](https://store.arduino.cc/products/mkr-sd-proto-shield)), or one of the common SD modules.
```

- Remove superfluous parameters from store URLs (I assume these were not added intentionally and they only clutter up the content)
- Correct name of "MKR SD Proto Shield"

- An SD card that fits your SD slot.
- A way to connect the SD card to your computer.
- A way to make connections to the SWD pins on your target Arduino board.

> For the [Nano 33 IoT](https://store.arduino.cc/products/arduino-nano-33-iot?_gl=1%2A80ta1j%2A_ga%2AMjEyMzQ2MjgwOC4xNjY1NjUyNTY3%2A_ga_NEXN8H46L5%2AMTY3MTYyNzMzMS4xNjEuMS4xNjcxNjMwNDIwLjAuMC4w) and the **MKR Boards** other than **MKR1000** ([tutorial](https://support.arduino.cc/hc/en-us/articles/360014905199-How-to-burn-the-bootloader-to-Arduino-MKR1000-using-a-programmer?queryID=ae55269235f9756d76c0b66bbb8ffa69)), we found it the easiest to use a 0.1" pitch 2x3 POGO adapter. You could also solder wires to the test points if you prefer. 

Reviewer: ```suggestion
For the [Nano 33 IoT](https://store.arduino.cc/products/arduino-nano-33-iot?_gl=1%2A80ta1j%2A_ga%2AMjEyMzQ2MjgwOC4xNjY1NjUyNTY3%2A_ga_NEXN8H46L5%2AMTY3MTYyNzMzMS4xNjEuMS4xNjcxNjMwNDIwLjAuMC4w) and the **MKR Boards** other than **MKR1000** ([tutorial](https://support.arduino.cc/hc/en-us/articles/360014905199-How-to-burn-the-bootloader-to-Arduino-MKR1000-using-a-programmer?queryID=ae55269235f9756d76c0b66bbb8ffa69)), we found it the easiest to use a 0.1" pitch 2x3 pogo adapter. You could also solder wires to the test points if you prefer. 
```

Correct case of "pogo" ([reference](https://en.wikipedia.org/wiki/Pogo_pin))

- Select **Sketch > Include Library > Manage  Libraries…** from the Arduino IDE menu.
- Wait for the index updates to finish.
- In the "**Filter your search…**” field type ‘Adafruit DAP library’.
- Press **Enter**.
- Click on “**Adafruit DAP library by Adafruit**”.
- Click the **Install** button.
- You may now get a dialog asking whether to install the library's dependencies. If so, click the "**Install All**" button.
- Wait for the installation to finish
- Close the **Library Manager**
- Select **File > Example > Adafruit DAP Library > samd21 > flash_MKR_bootloaders** from the Arduino IDE menu. (despite the “MKR” in the sketch name, this also supports the **Nano 33 IoT**).
- Select your **programmer** Arduino board from the IDE’s **Tools > Board** menu.
- Select the port of the **programmer** Arduino board from the Arduino IDE’s Tools > Port menu.
- Select **Sketch > Upload** from the Arduino IDE menu.
- Unplug the **programmer** Arduino board from your computer.
- Connect the **programmer** Arduino board to the target Arduino board as follows:

    | Programmer  | Target      |
    | ----------- | ----------- |
    | VCC         | +3V3        |

>     | 12          | SWDIO       |
>     | 11          | SWCLK       |
>     | GND         | GND         |
>     | 9           | RESETN      |

Reviewer: ```suggestion
    | 1           | SWDIO       |
    | 2           | SWCLK       |
    | GND         | GND         |
    | 0           | RESETN      |
```

Unfortunately they chose different pins to use in the two example sketches. This example uses pins 1, 2, 0:

https://github.com/adafruit/Adafruit_DAP/blob/1.8.0/examples/samd21/flash_MKR_bootloaders/flash_MKR_bootloaders.ino#L4-L6

    SWD pads on MKR boards other than MKR 1000:

    ![](assets/SWDpadsMKR.png)

    MKR1000 SWD header pinout:

    ![](assets/SWDMKR1000header.png)

    Nano 33 IoT SWD pads:

    ![](assets/SWDpadsNano33IoT.png)

- Plug the USB cable of the **programmer** Arduino board into your computer.
- Select **Tools > Serial Monitor** from the Arduino IDE’s menu.
- Select "**No line ending**” from the dropdown menu near the bottom right corner of the **Serial Monitor** window.
- You should now see some instructions for using the sketch in the **Serial Monitor** output field. 

> This includes a menu of the boards supported by the sketch:

Reviewer: ```suggestion
  This includes a menu of the boards supported by the sketch:
```

Indent sentence to make it part of associated bullet point

   `
  [...]

   Select Arduino MKR board to erase and flash with bootloader:
   
   Z     -> Arduino Zero (6504 bytes)
   MZ    -> Arduino MKR Zero (6408 bytes)
   1000  -> Arduino MKR 1000 WIFI (6408 bytes)
   1010  -> Arduino MKR WIFI 1010 (7984 bytes)

   [...]

>    `

Reviewer: Something is wrong with the markup here. Check how it is rendered:

https://docscontent-hannes7eicherbootloaderreloadi.gatsbyjs.io/tutorials/generic/bootloader-reloading#instructions-1

![image](https://user-images.githubusercontent.com/8572152/212474078-6e29e8e3-7317-4622-93cb-3cee3809a26d.png)

(note it is not in a code block as intended.

It might be caused by the inconsistent indentation of the text (the first line of the block is indented by two spaces, while the rest is indented three spaces.

I don't know about the Arduino documentation system, but from the standpoint of normal Markdown usage I find this markup very odd. The standard markup for code blocks is a "[fenced code block](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks)", where the block content is wrapped with triple backticks (` ``` `):

````markdown
  ```
  [...]

  Select Arduino MKR board to erase and flash with bootloader:

  Z     -> Arduino Zero (6504 bytes)
  MZ    -> Arduino MKR Zero (6408 bytes)
  1000  -> Arduino MKR 1000 WIFI (6408 bytes)
  1010  -> Arduino MKR WIFI 1010 (7984 bytes)

  [...]
  ```
````

Single backticks are normally only used for inline content.

Author: You're right and it should indeed be three.
